### ISSUE TITLE
Is it possible to exclude default values on message decode?

### ISSUE THREAD

**Issue by robtayl0r** [stephenh/ts-proto#627]
Is there a plan to support excluding missing fields on decode? The documentation about how to work around this seems to suggest using the toJSON method, but that currently doesn't have any effect on default values:

```
Foo.toJSON({ bar: undefined }); // => { }
Foo.toJSON({ bar: '' }); // => { bar: '' } 
```

It would be desirable to, within a service method, have the `call.request` message not decoded with the default fields "filled in".

```typescript
import * as pb from './generated`;

const ServerImpl: pb.ServiceNameServer = {
  save: async (call, callback) => {
    // current
    typeof call.request.stringNotSentByClient; // string, ""
    typeof call.request.intNotSentByClient; // number, 0

    // desired decoded message
    typeof call.request.stringNotSentByClient; // undefined
    typeof call.request.intNotSentByClient; // undefined
    typeof call.request.intSentByClientAsZero; // number, 0
  },
}
```

For reference, the `@grpc/proto-loader` package's type generation script drops the defaults/empty fields by default and has a `--defaults` option to set this.

**Comment by stephenh** [comment]
@robtayl0r afaiu this would be a bad idea b/c in proto3 when a client sends `someInt = 0`, proto3 says "well, 0 is the default value for ints anyway, I'm going to save wire cost and not send it".

So, when you decode `someInt`, and it's "not on the wire", you really don't know if the client "did not set it" or "set it to zero but the runtime left it off".

Obviously the `@grpc/proto-loader` guys are experts on this, but dunno, seems weird that is their default behavior, with my current understanding of things.

Let me know what I'm missing; fwiw ts-proto does have a long/sordid history of "adding yet another flag for new custom behavior", so I'm not 100% against a PR that would support this...but still seems weird, and I'd like to understand more what I'm missing and why `@grpc/proto-loader` considers this behavior as something worth supporting.

I almost wonder if it was a bug in their first release, and now they have to be backwards compatible; but again, lmk if it does actually make sense.

**Comment by robtayl0r** [comment]
Thanks for the reply. Zero is a definitely a problem. I was thinking one way around this would be to have a call to explicitly zero out a value, e.g. `Service.SetScoreToZero()`.

Maybe I'm thinking about things from the wrong angle, but the use case I'm dealing with is when a service call is responsible for updating an existing record/model/row, when the caller is only sending a delta of changes. 

Assuming we have a row in a users table like:
| id | name | age |
| - | - | - |
| "foo" | "bar" | 10 |

```typescript
const message: UpdateUserRequest = {
  id: 'foo'
  age: 11
};

serviceClient.updateUser(message, errback);

// on the server side
ServiceImpl = { 
  update: (call:  grpc.ServerUnaryCall<UpdateUserRequest, UpdateUserResponse>, callback) {
    // where call.request = { id : 'foo', name: '', age: 11 }
    someOrm.update(call.request.id, call.request);
    // we've now overwrote the name column with an empty string
  }
}
```

| id | name | age |
| - | - | - |
| "foo" | "" | 11 |

It seems as if the only approach is to always assume that the deserialized message is the complete representation of the related object, and to update the record as-is. So sending partial objects with just a subset of fields/columns we wish to update would not be supported. My feeling is that this would be a common use case.

**Comment by stephenh** [comment]
Yeah, partial update APIs with just native string/int primitive types are not possible in protobuf, fundamentally.

You can use either a) the well known types like StringValue or b) later releases of proto3 brought back the "optional" keyword like `optional int32 age = 3`.

ts-proto supports both of those; understood that both of those option require changing your schema, but that's just the reality of protobuf's primitive types, unfortunately.
